src/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.AddSub.cs

```diff
@@ -68,5 +128,208 @@ internal static partial class BigIntegerCalculator
+        }
+
+        [SecuritySafeCritical]
+        private unsafe static void SubtractSelf(uint* left, int leftLength,
+                                                uint* right, int rightLength)
+        {
+            Debug.Assert(leftLength >= 0);
+            Debug.Assert(rightLength >= 0);
+            Debug.Assert(leftLength >= rightLength);
+            Debug.Assert(Compare(left, leftLength, right, rightLength) >= 0);
+
+            // Executes the "grammar-school" algorithm for computing z = a - b.
+            // Same as above, but we're writing the result directly to a and
+            // stop execution, if we're out of b and c is already 0.
+
+            int i = 0;
+            long carry = 0L;
+
+            for (; i < rightLength; i++)
+            {
+                long digit = (left[i] + carry) - right[i];
+                left[i] = (uint)digit;
+                carry = digit >> 32;
+            }
+            for (; carry != 0 && i < leftLength; i++)
+            {
+                long digit = left[i] + carry;
+                left[i] = (uint)digit;
+                carry = digit >> 32;
+            }
+        }
+
+        public static int Compare(uint[] left, uint[] right)
+        {
+            Debug.Assert(left != null);
+            Debug.Assert(right != null);
+
+            int i = Math.Max(left.Length, right.Length) - 1;
+
+            for (; i >= left.Length; i--)
+            {
+                if (right[i] != 0)
+                    return -1;
+            }
+
+            for (; i >= right.Length; i--)
+            {
+                if (left[i] != 0)
+                    return 1;
+            }
+
+            for (; i >= 0; i--)
+            {
+                if (left[i] < right[i])
+                    return -1;
+                if (left[i] > right[i])
+                    return 1;
+            }
+
+            return 0;
+        }
+
+        [SecuritySafeCritical]
+        private unsafe static int Compare(uint* left, int leftLength,
+                                          uint* right, int rightLength)
+        {
+            Debug.Assert(leftLength >= 0);
+            Debug.Assert(rightLength >= 0);
+
+            int i = Math.Max(leftLength, rightLength) - 1;
+
+            for (; i >= leftLength; i--)
+            {
+                if (right[i] != 0)
+                    return -1;
+            }
+
+            for (; i >= rightLength; i--)
+            {
+                if (left[i] != 0)
+                    return 1;
+            }
+
+            for (; i >= 0; i--)
+            {
+                if (left[i] < right[i])
+                    return -1;
+                if (left[i] > right[i])
+                    return 1;
+            }
+
+            return 0;
+        }
+
+        private static int Length(uint[] value)
+        {
+            Debug.Assert(value != null);
+            Debug.Assert(value.Length >= 1);
+
+            // One leading zero seems to be OK (?)
```

Comment:
There are corner cases where the array can have one leading zero, which is quite annoying. Maybe eliminating those cases would be a more elegant solution? I could add an appropriate check to `BigInteger.AssertValid` and then try to get rid of them. Hence, we wouldn't need that code anymore...